Tensorflow Philosophy blogpost (#465)
huggingface/blog: tensorflow-philosophy.md

---
title: "Hugging Face's TensorFlow Philosophy"
thumbnail: /blog/assets/91_tf_xla_generate/thumbnail.png
---

<h1>
    Hugging Face's TensorFlow Philosophy
</h1>

<div class="blog-metadata">
    <small>Published Aug 12, 2022.</small>
    <a target="_blank" class="btn no-underline text-sm mb-5 font-sans" href="https://github.com/huggingface/blog/blob/main/tf_xla_generate.md">
        Update on GitHub
    </a>
</div>

<div class="author-card">
    <a href="/rocketknight1">
        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1660312628256-60ba519750effef3a58beac3.png?w=200&h=200&f=face">
        <div class="bfc">
            <code>rocketknight1</code>
            <span class="fullname">Matthew Carrigan</span>
        </div>
    </a>
</div>

### Introduction

> Despite increasing competition from PyTorch and JAX, TensorFlow remains [the most-used deep learning framework](https://twitter.com/fchollet/status/1478404084881190912?lang=en). It also differs from those other two libraries in some very important ways. In particular, it’s quite tightly integrated with its high-level API `Keras`, and its data loading library `tf.data`.

Reviewer: Lol, this sets the tone!

Author: It's revenge for the :bell: tweet

As refined, respected and highly attractive TensorFlow engineers, we want to use the incredible power and flexibility of cutting-edge models, but we want to handle them with the tools and API we’re familiar with. This blogpost will be about the choices we make at Hugging Face to enable that, and what to expect from the framework as a TensorFlow programmer.

### Interlude: 30 Seconds to :huggingface:

Experienced users can feel free to skim or skip this section, but if this is your first encounter with Hugging Face and `transformers`, I should start by giving you an overview of the core idea of the library: You just ask for a pretrained model by name, and you get it in one line of code. The easiest way is to just use the `TFAutoModel` class:

```py
from transformers import TFAutoModel

> model = TFAutoModel.from_pretrained('bert-base-cased')

Reviewer: ```suggestion
model = TFAutoModel.from_pretrained("bert-base-cased")
```

    # I have a truly wonderful idea for this
    # which this code box is too short to contain
```

#### Philosophy #2: Loss functions are provided by default, but can be easily changed.

In Keras, the standard way to train a model is to create it, then `compile()` it with an optimizer and loss function, and finally `fit()` it. It’s very easy to load a model with transformers, but setting the loss function can be tricky - even for standard language model training, your loss function can be surprisingly non-obvious, and some hybrid models have extremely complex losses.

Our solution to that is simple: If you `compile()` without a loss argument, we’ll give you the one you probably wanted. Specifically, we’ll give you one that matches both your base model and output type - if you `compile()` a BERT-based masked language model without a loss, we’ll give you a masked language modelling loss that handles padding and masking correctly, and will only compute losses on corrupted tokens, exactly matching the original BERT training process. If for some reason you really, really don’t want your model to be compiled with any loss at all, then simply specify `loss=None` when compiling.

```py

> model = TFAutoModelForQuestionAnswering.from_pretrained('bert-base-cased')
> model.compile(optimizer='adam')  # No loss argument!

Reviewer: ```suggestion
model = TFAutoModelForQuestionAnswering.from_pretrained("bert-base-cased")
model.compile(optimizer="adam")  # No loss argument!
```

#### Philosophy #4: You shouldn’t have to write your own data pipeline, especially for common tasks

In addition to `transformers`, a huge open repository of pre-trained models, there is also 🤗 `datasets`, a huge open repository of datasets - text, vision, audio and more. These datasets convert easily to TensorFlow Tensors and Numpy arrays, making it easy to use them as training data. Here’s a quick example showing us tokenizing a dataset and converting it to Numpy. As always, make sure your tokenizer matches the model you want to train with, or things will get very weird!

```py
from datasets import load_dataset
from transformers import AutoTokenizer, TFAutoModelForSequenceClassification
from tensorflow.keras.optimizers import Adam

dataset = load_dataset("glue", "cola")  # Simple text classification dataset
dataset = dataset["train"]  # Just take the training split for now

# Load our tokenizer and tokenize our data
tokenizer = AutoTokenizer.from_pretrained('bert-base-cased')
tokenized_data = tokenizer(dataset["text"], return_tensors="np", padding=True)
labels = np.array(dataset["label"]) # Label is already an array of 0 and 1

# Load and compile our model

> model = TFAutoModelForSequenceClassification.from_pretrained('bert-base-cased')

Reviewer: ```suggestion
model = TFAutoModelForSequenceClassification.from_pretrained("bert-base-cased")
```

)

model.fit(tf_dataset)
```
Why is [prepare_tf_dataset()](https://huggingface.co/docs/transformers/main/en/main_classes/model#transformers.TFPreTrainedModel.prepare_tf_dataset) a method on your model? Simple: Because your model knows which columns are valid as inputs, and automatically filters out columns in the dataset that aren't valid input names! If you’d rather have more precise control over the `tf.data.Dataset` being created, you can use the lower level [Dataset.to_tf_dataset()](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset.to_tf_dataset) instead.

#### Philosophy #5: XLA is great!

[XLA](https://www.tensorflow.org/xla) is the just-in-time compiler shared by TensorFlow and JAX. It converts linear algebra code into more optimized versions that run quicker and use less memory. It’s really cool and we try to make sure that we support it as much as possible. It’s extremely important for allowing models to be run on TPU, but it offers speed boosts for GPU and even CPU as well! To use it, simply `compile()` your model with the `jit_compile=True` argument (this works for all Keras models, not just Hugging Face ones):

```py

> model.compile(optimizer='adam', jit_compile=True)

Reviewer: ```suggestion
model.compile(optimizer="adam", jit_compile=True)
```

```py
# This is a new feature, so make sure to update to the latest version of transformers!
# You will also need to pip install tensorflow_text

import tensorflow as tf
from transformers import TFAutoModel, TFBertTokenizer

class EndToEndModel(tf.keras.Model):
    def __init__(self, checkpoint):
        super().__init__()
        self.tokenizer = TFBertTokenizer.from_pretrained(checkpoint)
        self.model = TFAutoModel.from_pretrained(checkpoint)

    def call(self, inputs):
        tokenized = self.tokenizer(inputs)
        return self.model(**tokenized)

model = EndToEndModel(checkpoint='bert-base-cased')

test_inputs = [

>     'This is a test sentence!',
>     'This is another one!'

Reviewer: ```suggestion
    "This is a test sentence!",
    "This is another one!",
```
